igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerService.groovy

```diff
@@ -0,0 +1,349 @@
+                //is not in ISO8601 format, and parsing fails. The following is a temporary
+                //workaround - the getRunById call below gets the same Run object but Wercker
+                //returns the date in the ISO8601 format for this case.
+                Run run = werckerClient.getRunById(authHeaderValue, runInfo.get('id'))
+
+                //Create an entry in the WerckerCache for this new run. This will also generate
+                //an integer build number for the run
+                Map<String, Integer> runIdBuildNumbers = cache.updateBuildNumbers(
+                        master, appAndPipelineName, Collections.singletonList(run))
+
+                log.info("Triggered run {} at URL {} with build number {}",
+                        kv("runId", run.id), kv("url", run.url),
+                        kv("buildNumber", runIdBuildNumbers.get(run.id)))
+
+                //return the integer build number for this run id
+                return runIdBuildNumbers.get(run.id)
+            } catch (RetrofitError e) {
+                def body = e.getResponse().getBody()
+                String wkrMsg
+                if (body instanceof TypedByteArray) {
+                    wkrMsg = new String(((retrofit.mime.TypedByteArray) body).getBytes())
+                } else {
+                    wkrMsg = body.in().text
+                }
+                log.error("Failed to trigger build for pipeline {}. {}", kv("pipelineName", pipelineName), kv("errMsg", wkrMsg))
+                throw new BuildController.BuildJobError(
+                "Failed to trigger build for pipeline ${pipelineName}! Error from Wercker is: ${wkrMsg}")
+            }
+        } else {
+            throw new BuildController.InvalidJobParameterException(
+            "Could not retrieve pipeline ${pipelineName} for application ${appName} from Wercker!")
+        }
+    }
+
+    /**
+     * Returns List of all Wercker jobs in the format of type/org/app/pipeline 
+     */
+    public List<String> getJobs() {
+        List<String> jobs = []
+        long start = System.currentTimeMillis();
+        List<Application> apps = getApplications();
+        start = System.currentTimeMillis()
+        apps.each { app ->
+            try {
+                List<Pipeline> pipelines = app.pipelines? app.pipelines : []; //getPipelines(app.owner.name, app.name);
+                jobs.addAll( pipelines.collect {
+                    it.type + SPLITOR + app.owner.name + SPLITOR + app.name + SPLITOR + it.name } )
+            } catch(retrofit.RetrofitError err) {
+                log.error "Error getting pipelines for ${app.owner.name } ${app.name} ${err}"
+            }
+        }
+        log.debug "getPipelines: ${jobs.size()} pipelines in ${System.currentTimeMillis() - start}ms"
+        return jobs
+    }
+
+    List<Application> getApplications() {
+        new SimpleHystrixCommand<List<Application>>(groupKey, buildCommandKey("getApplications"), {
```

Comment:
I'm curious why this (and the next) call to the `werckerClient` go through `Hystrix` but none of the other ones do.

Reply:
I think this was simply an artifact of copying the JenkinsService as a starting point. We didn't think about when to use it or not.
Looking at the JenkinsService now, I see that a number of the calls use Hystrix but a handful do not - these seem to be the ones that retrieve data on a single item (e.g. get build by build number), and the ones that do POSTs (start/stop a build etc)

Would your suggestion be to follow a similar pattern and use Hystrix for most of the http calls?

Reply:
In general, I think using Hystrix is a good idea for communicating with systems outside of Spinnaker's control (or even outside of the control of the particular microservice in question), as it can help avoid an outage in the external system spilling into Spinnaker, and allows graceful degradation (by supplying fallback functions).  So I'd probably recommend using it.

Reply:
all using Hystrix now.